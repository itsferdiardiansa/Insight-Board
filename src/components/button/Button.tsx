export default function Button() {
  return <button data-testid="btn">Click me!</button>
}
