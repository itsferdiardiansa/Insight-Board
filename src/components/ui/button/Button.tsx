'use client'

import { type ReactNode, ButtonHTMLAttributes, forwardRef } from 'react'
import { cva, type VariantProps } from 'class-variance-authority'
import { Slot, Slottable } from '@radix-ui/react-slot'
import { cn } from '@/utils/tailwind'
import { FaCircleNotch } from 'react-icons/fa'

const buttonSizes = {
  sm: 'px-2.5 py-1.5 text-sm',
  md: 'px-3.5 py-2.5 text-base',
  lg: 'px-4.5 py-3.5 text-lg',
}

const buttonVariants = cva(
  'inline-flex items-center justify-center gap-1.5 whitespace-nowrap font-bold tracking-wide cursor-pointer transition-all duration-300 ease-in-out',
  {
    variants: {
      variant: {
        primary: 'bg-(--primary) text-neutral-100 hover:bg-(--primary-hover)',
        outlinePrimary:
          'outline outline-(--primary) text-(--primary) hover:bg-(--primary) hover:text-neutral-100',
        secondary:
          'bg-(--secondary) text-neutral-100 hover:bg-(--secondary-hover)',
        outlineSecondary:
          'outline outline-(--secondary) hover:bg-(--secondary) hover:text-neutral-100',
        destructive:
          'bg-(--destructive) text-neutral-100 hover:bg-(--destructive-hover)',
        outlineDestructive:
          'outline outline-(--destructive) text-(--destructive) hover:bg-(--destructive) hover:text-neutral-100',
        light: 'bg-white text-neutral-800 hover:bg-gray-50',
        outlineLight:
          'outline outline-gray-300 text-neutral-800 hover:bg-white hover:text-black',
        ghost: 'text-(--secondary)',
      },
      size: buttonSizes,
      rounded: {
        none: 'rounded-none',
        sm: 'rounded-sm',
        md: 'rounded-md',
        full: 'rounded-full',
      },
      disabled: {
        true: 'bg-(--muted) hover:bg-(--muted) disabled:cursor-not-allowed disabled:opacity-50',
      },
      fullWidth: {
        true: 'w-full',
      },
    },
    compoundVariants: [
      { size: 'sm', rounded: 'full', className: 'px-2.5' },
      { size: 'md', rounded: 'full', className: 'px-3.5' },
      { size: 'lg', rounded: 'full', className: 'px-4.5' },

      { disabled: false, rounded: 'none', className: 'hover:rounded-md' },
      { disabled: false, rounded: 'sm', className: 'hover:rounded-lg' },
      { disabled: false, rounded: 'md', className: 'hover:rounded-xl' },
    ],
    defaultVariants: {
      variant: 'primary',
      // size: 'md',
      fullWidth: false,
      rounded: 'md',
    },
  }
)

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  icon?: ReactNode
  iconPosition?: 'left' | 'right'
  size?: keyof typeof buttonSizes
  disabled?: boolean
  loading?: boolean
  asChild?: boolean
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      children,
      icon,
      iconPosition = 'left',
      variant,
      size = 'md',
      fullWidth,
      rounded,
      loading = false,
      asChild = false,
      disabled,
      ...props
    },
    ref
  ) => {
    const Comp = asChild ? Slot : 'button'

    return (
      <Comp
        {...props}
        ref={ref}
        data-testid="btn"
        disabled={disabled || loading}
        className={cn(
          buttonVariants({ variant, fullWidth, rounded, disabled }),
          asChild && buttonSizes[size],
          iconPosition === 'right' && 'flex-row-reverse',
          className
        )}
      >
        {loading && <FaCircleNotch className="animate-spin" aria-hidden />}
        {!loading && (
          <Slottable>
            {asChild ? (
              children
            ) : (
              <div
                className={cn(
                  'flex gap-(--space-sm) items-center justify-center',
                  buttonSizes[size as keyof typeof buttonSizes]
                )}
              >
                {!loading && icon && (
                  <span className="contents self-stretch">{icon}</span>
                )}
                <span>{children}</span>
              </div>
            )}
          </Slottable>
        )}
      </Comp>
    )
  }
)

Button.displayName = 'Button'
