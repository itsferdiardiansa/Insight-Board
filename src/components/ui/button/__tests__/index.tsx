import { fireEvent, render, screen } from '@testing-library/react'
import { Button, type ButtonProps } from '..'
import { FaUser } from 'react-icons/fa'
import React from 'react'
import {} from '../Button'

const CustomLink = React.forwardRef<
  HTMLAnchorElement,
  React.AnchorHTMLAttributes<HTMLAnchorElement>
>((props, ref) => {
  return (
    <a
      ref={ref}
      {...props}
      className="text-violet-800"
      data-testid="custom-link"
    >
      Custom Link
    </a>
  )
})
CustomLink.displayName = 'CustomLink'

describe('components/Button', () => {
  it('renders contents correctly', () => {
    const element = render(<Button>Click me!</Button>)

    const btn = screen.queryByTestId('btn')

    expect(btn?.textContent).toEqual('Click me!')
    expect(element).toMatchSnapshot()
  })

  it('renders with default props', () => {
    render(<Button>Click Me</Button>)
    const button = screen.getByTestId('btn')
    expect(button).toBeInTheDocument()
    expect(button).toHaveTextContent('Click Me')
    expect(button).not.toBeDisabled()
  })

  it.each([
    ['primary'],
    ['secondary'],
    ['destructive'],
    ['outline'],
    ['outlinePrimary'],
    ['outlineSecondary'],
    ['outlineDestructive'],
    ['ghost'],
  ])('renders with variant %s', variant => {
    render(<Button variant={variant as ButtonProps['variant']}>Test</Button>)
    const button = screen.getByTestId('btn')
    expect(button).toBeInTheDocument()
    expect(button).toHaveTextContent('Test')
  })

  it.each([['sm'], ['md'], ['lg']])('applies size %s', size => {
    render(<Button size={size as ButtonProps['size']}>Sized</Button>)
    const button = screen.getByTestId('btn')
    expect(button).toBeInTheDocument()
  })

  it('applies fullWidth when true', () => {
    render(<Button fullWidth>Wide</Button>)
    const button = screen.getByTestId('btn')
    expect(button.className).toMatch(/w-full/)
  })

  it('renders with left icon', () => {
    render(
      <Button icon={<FaUser />} iconPosition="left">
        Profile
      </Button>
    )
    expect(screen.getByText('Profile')).toBeInTheDocument()
    expect(screen.getByTestId('btn').querySelector('svg')).toBeInTheDocument()
  })

  it('renders with right icon', () => {
    render(
      <Button icon={<FaUser />} iconPosition="right">
        Profile
      </Button>
    )
    expect(screen.getByText('Profile')).toBeInTheDocument()
    const button = screen.getByTestId('btn')
    expect(button.querySelector('svg')).toBeInTheDocument()
  })

  it('renders loading spinner', () => {
    render(<Button loading>Loading</Button>)
    const button = screen.getByTestId('btn')
    expect(button).toBeDisabled()
    expect(button.querySelector('svg')).toHaveClass('animate-spin')
    expect(button).not.toHaveTextContent('Loading') // because children is visually hidden
  })

  it('disables when `disabled` is passed', () => {
    render(<Button disabled>Blocked</Button>)
    const button = screen.getByTestId('btn')
    expect(button).toBeDisabled()
  })

  it('calls onClick when clicked', () => {
    const onClick = jest.fn()
    render(<Button onClick={onClick}>Click</Button>)
    fireEvent.click(screen.getByTestId('btn'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('renders as child if `asChild` is true', () => {
    const element = render(
      <Button asChild>
        <CustomLink />
      </Button>
    )

    const link = screen.getByTestId('custom-link')
    expect(link).toBeInTheDocument()
    expect(link).toHaveTextContent('Custom Link')
    expect(element).toMatchSnapshot()
  })
})
