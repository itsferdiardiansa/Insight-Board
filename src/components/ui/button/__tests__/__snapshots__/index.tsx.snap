// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Button renders as child if \`asChild\` is true 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="text-violet-800"
        data-testid="custom-link"
      >
        Custom Link
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="text-violet-800"
      data-testid="custom-link"
    >
      Custom Link
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`components/Button renders contents correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap font-semibold cursor-pointer transition-colors focus:outline-none disabled:cursor-not-allowed disabled:opacity-50 border-violet-800 bg-violet-800 text-white hover:bg-violet-700 text-base rounded-md px-4 py-2"
        data-testid="btn"
      >
        Click me!
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap font-semibold cursor-pointer transition-colors focus:outline-none disabled:cursor-not-allowed disabled:opacity-50 border-violet-800 bg-violet-800 text-white hover:bg-violet-700 text-base rounded-md px-4 py-2"
      data-testid="btn"
    >
      Click me!
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
